### PR TITLE
[DLSP21] 07-3

### PR DESCRIPTION
**Pull request by mdagar6**
07-3

### CODE REVIEW

**Comment by Atcold** [comment]
Please, run the website locally first.
Also, read the [instructions](https://atcold.github.io/pytorch-Deep-Learning/en/about/).

File: docs/en/week07/07-3.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+$$
+F(y) = C(y,\tilde{y}) + R(h)
+$$
+
+### Reconstruction Costs
+Below are the two examples of reconstruction energies:
+##### Real Valued Input:
+$$
+C(y,\tilde{y}) = ||y-\tilde{y}||^2 = ||y-Dec[Enc(y)]||^2
```
Comment (Atcold) [inline docs/en/week07/07-3.md:72]: The norm operator is `\Vert`.
**All vectors** should be encapsulated in `\vect{}`.

File: docs/en/week07/07-3.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+$$
+C(y,\tilde{y}) = - \sum_{i=1}^n{[y_ilog(\tilde{y_i}) + (1-y_i)log(1-\tilde{y_i})]}
+$$
+
+### Loss Function
+Average across all training samples of per sample loss function
+
+$$
+L(F(.),Y) = \frac{1}{m}\sum_{j=1}^m{l(F(.),y^{(j)})} \in R
```
Comment (Atcold) [inline docs/en/week07/07-3.md:87]: `\mathbb` perhaps?
`\cdot`.

File: docs/en/week07/07-3.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+$$
+We take the energy loss and try to push the energy down on $\tilde{y}$
+$$
+l_{energy}(F(.),y) = F(y)
```
Comment (Atcold) [inline docs/en/week07/07-3.md:91]: `\ell`.

File: docs/en/week07/07-3.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
+# Application of Autoencoders
+## DALL-E: Creating Images from Text
+
+A neural network that creates images from text captions. 
+![](https://i.imgur.com/FosonFM.jpg)
```
Comment (Atcold) [inline docs/en/week07/07-3.md:5]: Where's the caption?

**Review by Atcold — CHANGES_REQUESTED** [review]
More feedback.

File: docs/en/week07/07-3.md (hunk 4 above)
Comment (mdagar6) [reply docs/en/week07/07-3.md:14]: Text caption? It is in the image.

File: docs/en/week07/07-3.md (hunk 1 above)
Comment (mdagar6) [reply docs/en/week07/07-3.md]: Done

File: docs/en/week07/07-3.md (hunk 2 above)
Comment (mdagar6) [reply docs/en/week07/07-3.md]: Done

File: docs/en/week07/07-3.md (hunk 3 above)
Comment (mdagar6) [reply docs/en/week07/07-3.md]: Done

File: docs/en/week07/07-3.md (hunk 4 above)
Comment (Atcold) [reply docs/en/week07/07-3.md:13]: The figure's caption is missing.
**Fig 1:** description of the figure 1

Also, please, read [the instructions](https://atcold.github.io/pytorch-Deep-Learning/en/about/)!

File: docs/en/week07/07-3.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+Go to the [website](https://openai.com/blog/dall-e/) and play with the captions! 
+
+# Autoencoder
+Let's start with some definitions:
+
+## Definitions
+### Input
+
+$\boldsymbol{x}$: is observed during both training and testing
```
Comment (Atcold) [inline docs/en/week07/07-3.md:24]: I said to use `\vect{}` for vectors. It's a macro, so we handle how all vectors appear.
Similarly, use `\matr{}` for matrices.
Do read [the instructions](https://atcold.github.io/pytorch-Deep-Learning/en/about/) please.

File: docs/en/week07/07-3.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+## Autoencoder
+
+These kind of networks are used to learn the internal structure of an input and encode it in a hidden internal representation ($\boldsymbol{h}$) which expresses the input.
+
+We already learned how to train energy-based models, let's look at the below network:
+![](https://i.imgur.com/NAUmcJd.png)
+
+Here instead of computing the minimization of the energy $E$ with respect to $\boldsymbol{z}$, we use an encoder that approximates the minimization and provides a hidden representation $\boldsymbol{h}$ for a given $\boldsymbol{y}$.
+$$
+\boldsymbol{h} = Encoder(\boldsymbol{y})
```
Comment (Atcold) [inline docs/en/week07/07-3.md:48]: Use `\text{}` for unconventional function names. Here and everywhere else in this document.

File: docs/en/week07/07-3.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
+---
+lang-ref: ch.07-3
+title: Introduction to autoencoders
+lecturer: Alfredo Canziani
+authors: Vidit Bhargava, Monika Dagar
+date: 18 March 2021
+---
+# DALL-E: Creating Images from Text
+
+[![hackmd-github-sync-badge](https://hackmd.io/usp9PiUGRdGQzD93wE8Bsw/badge)](https://hackmd.io/usp9PiUGRdGQzD93wE8Bsw)
```
Comment (Atcold) [inline docs/en/week07/07-3.md:10]: What is this supposed to be?

File: docs/en/week07/07-3.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
+---
+lang-ref: ch.07-3
+title: Introduction to autoencoders
+lecturer: Alfredo Canziani
+authors: Vidit Bhargava, Monika Dagar
+date: 18 March 2021
+---
+# DALL-E: Creating Images from Text
```
Comment (Atcold) [inline docs/en/week07/07-3.md:8]: Why are there two titles?

image

Comment (Atcold) [reply docs/en/week07/07-3.md:8]: Perhaps use `##` for the subtitle?

File: docs/en/week07/07-3.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+$$
+C(\boldsymbol{y},\boldsymbol{\tilde{y}}) = \Vert \boldsymbol{y}-\boldsymbol{\tilde{y}} \Vert^2 =  \Vert \boldsymbol{y}-Dec[Enc(\boldsymbol{y})] \Vert^2
+$$
+This is the square ecludian distance between $\boldsymbol{y}$ and $\boldsymbol{\tilde{y}}$.
+
+##### Binary Input:
+In the case of binary input, we can simply use binary cross-entropy
+
+$$
+C(\boldsymbol{y},\boldsymbol{\tilde{y}}) = - \sum_{i=1}^n{[\boldsymbol{y_i}\log(\boldsymbol{\tilde{y_i}}) + (1-\boldsymbol{y_i})\log(1-\boldsymbol{\tilde{y_i}})]}
```
Comment (Atcold) [inline docs/en/week07/07-3.md:89]: No need of [] after the summation.

File: docs/en/week07/07-3.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+### Loss Function
+Average across all training samples of per sample loss function
+
+$$
+L(F(\cdot),\boldsymbol{Y}) = \frac{1}{m}\sum_{j=1}^m{\ell(F(\cdot),\boldsymbol{y}^{(j)})} \in \mathbb{R}
+$$
+We take the energy loss and try to push the energy down on $\boldsymbol{\tilde{y}}$
+$$
+\ell_{energy}(F(\cdot),\boldsymbol{y}) = F(\boldsymbol{y})
```
Comment (Atcold) [inline docs/en/week07/07-3.md:100]: Use `\text{}` for English.

File: docs/en/week07/07-3.md [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+$\boldsymbol{x}$: is observed during both training and testing 
+
+$\boldsymbol{y}$: is observed during training but not testing
+
+$\boldsymbol{z}$: is not observed (neither during training nor during testing).
+
+### Output
+$\boldsymbol{h}$: is computed from the input (hidden/internal)
+
+$\boldsymbol{\tilde{y}}$: is computed from the hidden (predicted $\boldsymbol{y}$, ~ means $circa$)
```
Comment (Atcold) [inline docs/en/week07/07-3.md:33]: *circa* is Latin, not math. Remove `$` and use `*` instead.

File: docs/en/week07/07-3.md [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+<center>
+<img src="{{site.baseurl}}/images/week07/07-3/DenoisingAutoEncoder.png" style="background-color:#DCDCDC;" /><br>
+<b>Figure 4:</b> Denoising Autoencoder Network architecture.
+</center>
+<!-- ![](https://i.imgur.com/WVcDLns.png) -->
+An important note: The noise added to the original input should be similar to what we expect in reality, so the model can easily recover from it.
+
+<center>
+<img src="{{site.baseurl}}/images/week07/07-3/DAEOutput.jpeg" style="background-color:#DCDCDC;" /><br>
+<b>Figure 5:</b> Network output.
```
Comment (Atcold) [inline docs/en/week07/07-3.md:127]: Use newer images from [this year's slides](https://drive.google.com/drive/u/1/folders/1cy7BHpZJQhkhbqwLShfq8UyYma11gGQa).

Comment (Atcold) [reply docs/en/week07/07-3.md:127]: Also, that's not quite the network's output. Please, describe the figure.

File: docs/en/week07/07-3.md [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+This is the square ecludian distance between $\boldsymbol{y}$ and $\boldsymbol{\tilde{y}}$.
+
+##### Binary Input:
+In the case of binary input, we can simply use binary cross-entropy
+
+$$
+C(\boldsymbol{y},\boldsymbol{\tilde{y}}) = - \sum_{i=1}^n{[\boldsymbol{y_i}\log(\boldsymbol{\tilde{y_i}}) + (1-\boldsymbol{y_i})\log(1-\boldsymbol{\tilde{y_i}})]}
+$$
+
+### Loss Function
```
Comment (Atcold) [inline docs/en/week07/07-3.md:92]: These are loss **functionals**, not *functions*.

File: docs/en/week07/07-3.md [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+Average across all training samples of per sample loss function
+
+$$
+L(F(\cdot),\boldsymbol{Y}) = \frac{1}{m}\sum_{j=1}^m{\ell(F(\cdot),\boldsymbol{y}^{(j)})} \in \mathbb{R}
```
Comment (Atcold) [inline docs/en/week07/07-3.md:96]: You forgot `\mathcal{}`.

**Review by Atcold — CHANGES_REQUESTED** [review]
Three figures are missing.
See comments as well.

File: docs/en/week07/07-3.md (hunk 9 above)
Comment (mdagar6) [reply docs/en/week07/07-3.md:89]: hmm I am not sure about this. Why?

File: docs/en/week07/07-3.md (hunk 12 above)
Comment (vidit23) [reply docs/en/week07/07-3.md]: Done

File: docs/en/week07/07-3.md (hunk 7 above)
Comment (vidit23) [reply docs/en/week07/07-3.md]: Removed

File: docs/en/week07/07-3.md [hunk 15]
```diff
@@ -0,0 +1,143 @@
+---
+lang-ref: ch.07-3
+title: Introduction to autoencoders
+lecturer: Alfredo Canziani
+authors: Vidit Bhargava, Monika Dagar
+date: 18 March 2021
+---
+$\newcommand{\vert}{\boldsymbol}$
```
Comment (Atcold) [inline docs/en/week07/07-3.md:8]: What is this???
`\vert` is the absolute value operator.
`\vect{}` is a shortcut for `\boldsymbol{}` already defined in the [preamble](https://github.com/Atcold/NYU-DLSP21/blob/master/docs/_layouts/custom.html#L7).

File: docs/en/week07/07-3.md [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+$\newcommand{\matr}{\boldsymbol}$
```
Comment (Atcold) [inline docs/en/week07/07-3.md:9]: Plz, remove both lines.

File: docs/en/week07/07-3.md [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+$\vert{x}$: is observed during both training and testing 
+
+$\vert{y}$: is observed during training but not testing
+
+$\vert{z}$: is not observed (neither during training nor during testing).
+
+#### Output
+$\vert{h}$: is computed from the input (hidden/internal)
+
+$\vert{\tilde{y}}$: is computed from the hidden (predicted $\vert{y}$, ~ means *circa*)
```
Comment (Atcold) [inline docs/en/week07/07-3.md:36]: image

Did you actually run the website locally?

Comment (Atcold) [reply docs/en/week07/07-3.md:36]: Just use `\vect{}` instead of the absolute value `\vert`.

**Review by Atcold — CHANGES_REQUESTED** [review]
Can you actually make sure what you submit is actually correct?
Thank you.

**Comment by Atcold** [comment]
Missing space before the equations won't make them displayed properly, as stated in the instruction.
image

Broken math.
image

**Comment by vidit23** [comment]
Hi Alf, sorry for the delay. Could you please review again?
The equation wasn't working even after adding lines because the symbol usage was incorrect. I ran the website and it seems to be working fine

**Comment by Atcold** [comment]
Looks good to me, thanks!

**Pull request merged**